src/storage/gc_worker.rs

```diff
@@ -0,0 +1,305 @@
+        }.map_err(Error::from)
+    }
+
+    /// Scan keys in the region. Returns scanned keys if any, and a key indicating scan progress
+    fn scan_keys(
+        &mut self,
+        ctx: &mut Context,
+        safe_point: u64,
+        from: Option<Key>,
+        limit: usize,
+    ) -> Result<(Vec<Key>, Option<Key>)> {
+        let snapshot = self.get_snapshot(ctx)?;
+        let mut reader = MvccReader::new(
+            snapshot,
+            Some(ScanMode::Forward),
+            !ctx.get_not_fill_cache(),
+            None,
+            None,
+            ctx.get_isolation_level(),
+        );
+
+        // If from.is_some(), it must not be the first scan of the region.
+        // So we must continue doing GC.
+        let skip_gc = from.is_none() && !reader.need_gc(safe_point, self.ratio_threshold);
+        let res = if skip_gc {
+            KV_GC_SKIPPED_COUNTER.inc();
+            Ok((vec![], None))
+        } else {
+            reader
+                .scan_keys(from.clone(), limit)
+                .map_err(Error::from)
+                .and_then(|(keys, next)| {
+                    if keys.is_empty() {
+                        assert!(next.is_none());
+                        if from.is_none() {
+                            KV_GC_EMPTY_RANGE_COUNTER.inc();
+                        }
+                    }
+                    Ok((keys, next))
+                })
+        };
+        self.stats.add_statistics(reader.get_statistics());
+        res
+    }
+
+    /// Clean up outdated data.
+    fn gc_keys(
+        &mut self,
+        ctx: &mut Context,
+        safe_point: u64,
+        keys: Vec<Key>,
+        mut next_scan_key: Option<Key>,
+    ) -> Result<Option<Key>> {
+        let snapshot = self.get_snapshot(ctx)?;
+        let mut txn = MvccTxn::new(
+            snapshot,
+            0,
+            Some(ScanMode::Forward),
+            ctx.get_isolation_level(),
+            !ctx.get_not_fill_cache(),
+        );
+        for k in keys {
+            // TODO: Duplicated code in scheduler.rs
+            let gc_info = txn.gc(&k, safe_point)?;
+
+            if gc_info.found_versions >= GC_LOG_FOUND_VERSION_THRESHOLD {
+                info!(
+                    "[region {}] GC found at least {} versions for key {}",
+                    ctx.get_region_id(),
+                    gc_info.found_versions,
+                    k
+                );
+            }
+            // TODO: we may delete only part of the versions in a batch, which may not beyond
+            // the logging threshold `GC_LOG_DELETED_VERSION_THRESHOLD`.
+            if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD {
+                info!(
+                    "[region {}] GC deleted {} versions for key {}",
+                    ctx.get_region_id(),
+                    gc_info.deleted_versions,
+                    k
+                );
+            }
+
+            if txn.write_size() >= MAX_TXN_WRITE_SIZE {
```

Comment:
I think it's better use the `gc_info` to indicate if current key has been fully handled.